java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java

```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
+                new HashSet<BVirtualMarker>(),
+                graph,
+                new Consumer<JsonObject>(){
+                    public void accept(JsonObject op) {
+                        if(jstring(op, "kind").equals("com.ibm.streamsx.topology.functional.java::HashAdder")){
+                            hashAdders.add(op);
+                        }
+                    }
+        });
+
+        for(JsonObject hashAdder : hashAdders){
+            relocateHashAdder(hashAdder);
+        }
+    }
+
+    private void relocateHashAdder(JsonObject hashAdder){
+        // $Unparallel$ -> hashAdder -> $Parallel$ -> hashRemover
+        Set<JsonObject> parents = GraphUtilities.getUpstream(hashAdder, graph);
+        
+        // hashAdder has multiple parents
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // hashAdder's parent has multiple children
+        if (GraphUtilities.getDownstream(parent, graph).size() != 1)
+            return;
+
+        // Only optimize the case where $Unparallel has only one child, and
+        // $parallel has only one parent
```

Comment:
See, again "$parallel has only one parent" but that's not what the code is checking for, so is the code correct and the comment wrong, or vice-versa?

Reply:
That's probably the comment I was looking for (assuming it is corrected), but it should be before the checks so that a reader understands the checks as they read down.